relay-server/src/services/project.rs

```diff
@@ -932,9 +933,39 @@ impl Project {
             Ok(current_limits.check_with_quotas(quotas, item_scoping))
         });
 
+        let check_nested_spans = state
+            .as_ref()
+            .is_some_and(|s| s.has_feature(Feature::ExtractSpansFromEvent));
+
+        #[derive(Debug, Deserialize)]
+        struct PartialEvent {
+            spans: SeqCount,
+        }
+
+        // We compute the number of nested spans before rate limiting since we will remove the
+        // transaction item in case it is rate limited.
+        let spans = if check_nested_spans {
+            envelope
+                .envelope()
+                .items()
+                .find(|i| *i.ty() == ItemType::Transaction && !i.spans_extracted())
+                .and_then(|i| serde_json::from_slice::<PartialEvent>(&i.payload()).ok())
+                // We do + 1, since we count the transaction itself because it will be extracted
+                // as a span and counted during the slow path of rate limiting.
+                .map_or(0, |p| p.spans.0 + 1)
+        } else {
+            0
+        };
+
         let (enforcement, mut rate_limits) =
             envelope_limiter.enforce(envelope.envelope_mut(), &scoping)?;
+
+        if check_nested_spans {
+            track_nested_spans_outcomes(&envelope, &enforcement, spans);
```

Comment:
I think we should integrate the new logic a little bit more with the `Enforcement`, to prevent having multiple code paths that do similar things. It's a good call to not inject the span count into the envelope summary, but it could still be part of the `Enforcement`. That is:

1. Pass a flag into `enforce` to conditionally check for spans in the transaction item and set `Enforcement::spans`.
2. `Enforcement::track_outcomes` automatically does the right thing

Reply:
We could possibly pass the managed envelope and some options to the `track_outcomes` and do it there, I preferred this at the moment, because `track_outcomes` already does work correctly for rate limits with the envelope. Modifying it will affect both cached/non-cached rate limits.

Having `track_outcomes` inspect the (managed) envelope seems like a footgun to me, since it is potentially built from completely different values than what the envelope contains (`assume_event`).